code/modules/urist/modules/shipbattles/overmapships/components/component_weapons.dm

```diff
@@ -5,133 +5,160 @@
 	var/ready = TRUE
-	var/obj/item/projectile/projectile_type
+	var/atom/movable/projectile_type
 	var/obj/machinery/shipweapons/weapon_type = /obj/machinery/shipweapons/beam/lightlaser
 
 /datum/shipcomponents/weapons/DoActivate()
 	if(ready && !broken)
 		Fire()
 
-/datum/shipcomponents/weapons/proc/Fire(obj/effect/urist/projectile_landmark/ship/map_landmark) //now infinitely less hacky and broken
-	if(broken) //check one more time, just in case. doing it this way can stop burst fire mid burst
+/datum/shipcomponents/weapons/proc/Fire(var/turf/start_turf, var/turf/target_turf, var/target_edge, var/target_z) //now infinitely less hacky and broken
+	if(broken || !projectile_type) //check one more time, just in case. doing it this way can stop salvo fire mid salvo
 		return
 
-	else
-		if(!map_landmark)
-			var/obj/effect/overmap/visitable/ship/combat/T = mastership.target_ship
-			map_landmark = pick(T.landmarks)
-
-		if(burst)
-			if(map_landmark)
-				ready = FALSE
-				if(shot_number < burst)
-					shot_number ++
-					map_landmark.Fire(projectile_type)
-					spawn(burst_delay)
-						Fire(map_landmark) //doing this so we fire from the same landmark. this gets us sustained fire on parts of the ship instead of the random scatter we have now.
-
-				else if(shot_number >= burst)
-					spawn(firedelay)
-						shot_number = 0
-						ready = TRUE
-		else
-			if(map_landmark)
-				map_landmark.Fire(projectile_type)
-				ready = FALSE
-				spawn(firedelay)
-					ready = TRUE
+	var/obj/effect/overmap/visitable/ship/combat/target = mastership?.target_ship
+	if(!target)
+		return
+
+	if(!start_turf)
+		if(!target_z)
+			target_z = pick(target.target_zs)
 
+		if(!target_edge)
+			if(target.target_dirs)
+				target_edge = pick(target.target_dirs)
+			else
+				target_edge = pick(GLOB.cardinal)
 
+		start_turf = spaceDebrisStartLoc(target_edge, target_z)
+
+	if(!target_turf && target_z)
+		target_turf = locate_target(target_z)
+
+	if(start_turf && target_turf)
+		ready = FALSE
+		if(shot_number < salvo)
+			shot_number ++
+			launch_atom(projectile_type, start_turf, target_turf)
+//			message_admins("An enemy ship has fired a [projectile_type] at the ship targeting <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[target_turf.x];Y=[target_turf.y];Z=[target_turf.z]'>(JMP)</a>.", 1)
+			if(shot_number >= salvo)
+				spawn(firedelay)
+					shot_number = 0
+					ready = TRUE
+			else
+				spawn(salvo_delay)
+					Fire(start_turf, null, target_edge, target_z) //doing this so we fire from the same spot at different targets.
+																//this gets us sustained fire on parts of the ship instead of a straight line of rockets, or the random scatter we had in the distant past.
 
+/datum/shipcomponents/weapons/proc/locate_target(var/target_z)
+	var/target_x = rand(mastership.target_ship.target_x_bounds[1],mastership.target_ship.target_x_bounds[2])
+	var/target_y = rand(mastership.target_ship.target_y_bounds[1],mastership.target_ship.target_y_bounds[2])
```

Comment:
This is currently uniform random; if you want Gaussian noise to simulate aiming at the center with random error instead, I can get you a sampler implementation.

Reply:
I think it's alright for the purposes of this. having random start locations for projectiles now makes things feel even more random anyways. Might be something worth looking into in the future though.